### ISSUE TITLE
Dead code elimination does not handle simple case

### ISSUE THREAD

**Issue by konsoletyper** [google/closure-compiler#2100]
Consider simple code:

``` javascript
"use strict";
(function() {
    var module$library = {};
    var module$main = {};

    (function(exports) {
        function foo() {
            return "foo";
        }

        function bar() {
            return "bar";
        }

        exports.foo = foo;
        exports.bar = bar;
    })(module$library);

    (function(exports, library) {
        function baz() {
            console.log(library.foo());
        }
        baz();
        exports.baz = baz;
    })(module$main, module$library);
})();
```

GCC with advanced mode enables produces the following code:

``` javascript
(function(){var a={};(function(b){b.c=function(){return"foo"};b.a=function(){return"bar"}})(a);(function(b){function a(){console.log("foo")}a();b.b=a})({},a)})();
```

It's expected that functions `foo` and `bar` are eliminated. However, they don't. Actually, this code can be simplified to the form:

``` javascript
console.log("foo");
```

**Comment by ChadKillingsworth** [comment]
That code is anything but simple. It creates aliases and requires function inlining to remove IIFE wrappers.

For CommonJS and ES6 modules, [we go to great lengths](https://github.com/google/closure-compiler/pull/2085) to alleviate these hurdles.

**Comment by konsoletyper** [comment]
I can't understand why it's untrivial to handle this case. Isn't it easy to inline these functions? What aliases does this code create? Do you mean `exports`? I guess inlining should replace all occurences of `exports` by `module$library` and `module$main` respectively.

**Comment by ChadKillingsworth** [comment]
Correct - if you don't create aliases with your IIFE wrappers, then you get the desired result:

http://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540formatting%2520pretty_print%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%2522use%2520strict%2522%253B%250A(function()%2520%257B%250A%2520%2520%2520%2520%252F**%2520%2540const%2520*%252F%250A%2520%2520%2520%2520var%2520module%2524library%2520%253D%2520%257B%257D%253B%250A%2520%2520%2520%2520%252F**%2520%2540const%2520*%252F%250A%2520%2520%2520%2520var%2520module%2524main%2520%253D%2520%257B%257D%253B%250A%250A%2520%2520%2520%2520(function()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520function%2520foo()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520return%2520%2522foo%2522%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%250A%2520%2520%2520%2520%2520%2520%2520%2520function%2520bar()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520return%2520%2522bar%2522%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%250A%2520%2520%2520%2520%2520%2520%2520%2520module%2524library.foo%2520%253D%2520foo%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520module%2524library.bar%2520%253D%2520bar%253B%250A%2520%2520%2520%2520%257D)()%253B%250A%250A%2520%2520%2520%2520(function()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520function%2520baz()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520console.log(module%2524library.foo())%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%2520%2520baz()%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%252F**%2520%2540const%2520*%252F%250A%2520%2520%2520%2520%2520%2520%2520%2520module%2524main.baz%2520%253D%2520baz%253B%250A%2520%2520%2520%2520%257D)()%253B%250A%257D)()%253B

**Comment by konsoletyper** [comment]
Why does not closure compiler eliminate these aliases?

**Comment by ChadKillingsworth** [comment]
Because it's hard. You can only eliminate the aliases in a small set of circumstances. For instance, both of the following cases cannot be inlined:

```js
var foo = 4;
(function(bar) { bar++; }(foo);
console.log(foo); // logs 4

var baz = {};
(function(bar) { bar = {}; bar.foo = 'foo'; })(baz);
console.log(baz); // logs "{}" - no property was added
```

You have to know that the alias was not the LHS of an expression, was a type that is passed by reference and the IIFE doesn't use the `arguments` keyword. That's probably not a complete list.

**Comment by konsoletyper** [comment]
Heh, this should be a couple of visitors. It's possible to build graph of trivial assignments like these:

1. `var x = y`
2. `foo(y)`; where foo is declared like `function foo(x) { ... }`
3. `x = foo(...)` where the last statement of `foo` is `return y`.

(all these examples yield `y -> x` edge)

And there should be defined "unsafe" operations over these variables. When a variable `v` encounters a unsafe operation (like reassignment or passing to function that uses `arguments`) then graph is traversed beginning from `v` and all reached nodes marked. All nodes of graph that are not marked are "safe" and can be optimized the way I'd like to.

What's a really big problem is to get "the complete list" of things that make variables unsafe. Do you have any idea of how to investigate this? Do you use some large code base to test GCC and find such cases.

**Comment by ChadKillingsworth** [comment]
Inlining functions is well understood and handled by the compiler. There isn't much work to be added there. The call graphs you describe are already being used.

Optimizing parameters for named functions is also already handled. The comments on the pass are very instructive: https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/OptimizeParameters.java

Optimizing parameters for IIFEs are a special case. They aren't optimized now and are actually easier. You know the function is only called once and the exact arguments are easy to look up (they are right there with it). If an IIFE argument can be inlined to the call argument, then function inlining takes over automatically.

Inlining IIFE arguments is safe under the following conditions:

 * The IIFE does not reference `arguments`
 * Only properties of the argument are accessed

> Do you use some large code base to test GCC and find such cases.

 Only every Google product

The big question is: does this pattern occur enough in real world code to justify a new pass?

**Comment by ChadKillingsworth** [comment]
I may have found an easy win. I think this rewriting is safe: http://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540formatting%2520pretty_print%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%2522use%2520strict%2522%253B%250A(function()%2520%257B%250A%2520%2520%2520%2520var%2520module%2524library%2520%253D%2520%257B%257D%253B%250A%2520%2520%2520%2520var%2520module%2524main%2520%253D%2520%257B%257D%253B%250A%250A%2520%2520%2520%2520(function()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520var%2520exports%2520%253D%2520module%2524library%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520function%2520foo()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520return%2520%2522foo%2522%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%250A%2520%2520%2520%2520%2520%2520%2520%2520function%2520bar()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520return%2520%2522bar%2522%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%250A%2520%2520%2520%2520%2520%2520%2520%2520exports.foo%2520%253D%2520foo%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520exports.bar%2520%253D%2520bar%253B%250A%2520%2520%2520%2520%257D)()%253B%250A%250A%2520%2520%2520%2520(function()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520var%2520exports%2520%253D%2520module%2524main%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520var%2520library%2520%253D%2520module%2524library%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520function%2520baz()%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520console.log(library.foo())%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520%257D%250A%2520%2520%2520%2520%2520%2520%2520%2520baz()%253B%250A%2520%2520%2520%2520%2520%2520%2520%2520exports.baz%2520%253D%2520baz%253B%250A%2520%2520%2520%2520%257D)()%253B%250A%257D)()%253B

**Comment by ChadKillingsworth** [comment]
Fixed via 9845959
